### PR TITLE
[10.0] porting stock_picking_invoice_link

### PR DESCRIPTION
**Pull request by eLBati**
based on 9.0 branch as https://github.com/OCA/stock-logistics-workflow/pull/296 is merged

### CODE REVIEW

File: stock_picking_invoice_link/views/account_invoice_view.xml [hunk 1]
```diff
@@ -0,0 +1,23 @@
+<?xml version="1.0" encoding="utf-8"?>
+<odoo>
+    <!--
+        Stock Picking Invoice Link module for Odoo
+        Copyright (C) 2013-2014 Akretion (http://www.akretion.com/)
+        @author: Alexis de Lattre <alexis.delattre@akretion.com>
+        The licence is in the file __openerp__.py
```
Comment (jesusVMayor) [inline stock_picking_invoice_link/views/account_invoice_view.xml]: change openerp to manifest

File: stock_picking_invoice_link/models/stock_move.py [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+# License AGPL-3 - See http://www.gnu.org/licenses/agpl-3.0.html
+
+from odoo import api, fields, models
+
+
+class StockMove(models.Model):
+    _inherit = "stock.move"
+
+    # Provide this field for backwards compatibility
+    invoice_line_ids = fields.Many2many(
```
Comment (jesusVMayor) [inline stock_picking_invoice_link/models/stock_move.py:14]: Why this field is still needed?

**Review by jesusVMayor — CHANGES_REQUESTED** [review]

File: stock_picking_invoice_link/views/account_invoice_view.xml (hunk 1 above)
Comment (pedrobaeza) [reply stock_picking_invoice_link/views/account_invoice_view.xml]: Even better if this adjusts to current template: https://github.com/OCA/maintainer-tools/blob/master/template/module/views/model_name_view.xml#L2

File: stock_picking_invoice_link/__init__.py [hunk 3]
```diff
@@ -1,21 +1,5 @@
...
-#    This program is distributed in the hope that it will be useful,
-#    but WITHOUT ANY WARRANTY; without even the implied warranty of
-#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
-#    GNU Affero General Public License for more details.
-#
-#    You should have received a copy of the GNU Affero General Public License
-#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
-#
-##############################################################################
+# © 2013-15 Agile Business Group sagl (<http://www.agilebg.com>)
```
Comment (chienandalu) [inline stock_picking_invoice_link/__init__.py]: Better keep init files minimal; no need for copyright

File: stock_picking_invoice_link/__manifest__.py [hunk 4]
```diff
@@ -1,37 +1,25 @@
...
-#    This program is distributed in the hope that it will be useful,
-#    but WITHOUT ANY WARRANTY; without even the implied warranty of
-#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
-#    GNU Affero General Public License for more details.
-#
-#    You should have received a copy of the GNU Affero General Public License
-#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
-#
-##############################################################################
+# © 2013-17 Agile Business Group (<http://www.agilebg.com>)
```
Comment (chienandalu) [inline stock_picking_invoice_link/__manifest__.py]: 'Copyright' instead of '©'

File: stock_picking_invoice_link/__manifest__.py [hunk 5]
```diff
@@ -1,37 +1,25 @@
...
+# © 2016 AvanzOSC (<http://www.avanzosc.es>)
+# © 2016 Pedro M. Baeza <pedro.baeza@tecnativa.com>
+# © 2017 Jacques-Etienne Baudoux <je@bcim.be>
+# License AGPL-3 - See http://www.gnu.org/licenses/agpl-3.0.html
 
 {
     'name': 'Stock Picking Invoice Link',
-    'version': '8.0.1.0.0',
+    'version': '10.0.0.2.0',
```
Comment (chienandalu) [inline stock_picking_invoice_link/__manifest__.py]: Why this version number? Follow conventions: https://github.com/OCA/maintainer-tools/blob/master/CONTRIBUTING.md#version-numbers

File: stock_picking_invoice_link/models/sale_order.py [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+from odoo import api, models
+
+
+class SaleOrderLine(models.Model):
+    _inherit = "sale.order.line"
+
+    @api.multi
+    def invoice_line_create(self, invoice_id, qty):
+        self.mapped('procurement_ids') \
+            .mapped('move_ids') \
```
Comment (chienandalu) [inline stock_picking_invoice_link/models/sale_order.py]: Why add unnecessary line breaks if they fit in the same line. This line can be styled with no need of backslash break.

File: stock_picking_invoice_link/models/stock_picking.py [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+
+
+class StockPicking(models.Model):
+    _inherit = "stock.picking"
+
+    invoice_ids = fields.Many2many(
+        comodel_name='account.invoice', copy=False, string='Invoices',
+        readonly=True)
+    # Provide this field for backwards compatibility
+    invoice_id = fields.Many2one(
```
Comment (chienandalu) [inline stock_picking_invoice_link/models/stock_picking.py:18]: Is this still needed?

**Review by chienandalu — CHANGES_REQUESTED** [review]
Just some styling comments until runbot is rebuilt for functional review.

**Comment by cubells** [comment]
@eLBati 

Can I add some commits to finish this pr?

File: stock_picking_invoice_link/__manifest__.py (hunk 5 above)
Comment (eLBati) [reply stock_picking_invoice_link/__manifest__.py]: @chienandalu see https://github.com/OCA/multi-company/pull/57#discussion_r136263566

**Comment by eLBati** [comment]
@cubells you are welcome

**Comment by MiquelRForgeFlow** [comment]
May you add commit 7ef2ce18d7e7eca4015dafc34838f6fd53e0e510 (https://github.com/OCA/stock-logistics-workflow/pull/357) and commits of https://github.com/OCA/stock-logistics-workflow/pull/339 🙂

**Comment by cubells** [comment]
@mreficent sure!

**Comment by cubells** [comment]
@eLBati 

I can't send my pr to your repo. Is It possible?

**Comment by pedrobaeza** [comment]
I will do it if you want.

**Comment by pedrobaeza** [comment]
I have integrated @cubells's PR.

@chienandalu @luismontalba please review.

File: stock_picking_invoice_link/README.rst [hunk 8]
```diff
@@ -1,43 +1,73 @@
...
+
+This module is also useful if you want to present data on the invoice report
+grouped by deliveries.
+
+Note that the links are only for products with an invoicing policy set on
+delivery.
+
+Usage
+=====
+
```
Comment (chienandalu) [inline stock_picking_invoice_link/README.rst:25]: Some usage instructions would be welcome

File: stock_picking_invoice_link/views/stock_view.xml [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+        <field name="inherit_id" ref="stock.view_picking_form"/>
+        <field name="arch" type="xml">
+            <field name="state" position="before">
+                    <button name="action_view_invoice" type="object"
+                        string="View Invoice" class="oe_highlight"
+                        attrs="{'invisible': [('invoice_ids', '=', [])]}"
+                        groups="base.group_user"/>
+            </field>
+            <field name="picking_type_id" position="after">
+                <field name="invoice_ids" string="Invoice"
```
Comment (chienandalu) [inline stock_picking_invoice_link/views/stock_view.xml:17]: Wouldn't a tree view be more practical?

File: stock_picking_invoice_link/views/stock_view.xml [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+    </record>
+
+    <!-- stock.move -->
+    <record id="view_move_form" model="ir.ui.view">
+        <field name="name">stock_picking_invoice_link.stock.move.form</field>
+        <field name="model">stock.move</field>
+        <field name="inherit_id" ref="stock.view_move_form"/>
+        <field name="arch" type="xml">
+            <field name="name" position="after">
+                <field name="invoice_line_ids" widget="many2many_tags" groups="account.group_account_invoice"/>
```
Comment (chienandalu) [inline stock_picking_invoice_link/views/stock_view.xml:33]: Same here, why not a tree view with more info

**Review by chienandalu — CHANGES_REQUESTED** [review]
There are lint errors. Some minor comments also:

File: stock_picking_invoice_link/views/stock_view.xml (hunk 9 above)
Comment (cubells) [reply stock_picking_invoice_link/views/stock_view.xml:17]: I think not. View adds two items to show invoice information in picks:
- a button to open invoice/s directly
- a field to quickly see if picking is invoiced.
I think is enough information:
seleccio_329

File: stock_picking_invoice_link/views/stock_view.xml (hunk 10 above)
Comment (cubells) [reply stock_picking_invoice_link/views/stock_view.xml:33]: I think default information is enough:

seleccio_330

Comment (chienandalu) [reply stock_picking_invoice_link/views/stock_view.xml:33]: But that's the invoice form view not the stock move form view

File: stock_picking_invoice_link/views/stock_view.xml (hunk 9 above)
Comment (chienandalu) [reply stock_picking_invoice_link/views/stock_view.xml:17]: Oh, I didn't notice the button, sorry

**Comment by cubells** [comment]
@pedrobaeza 

I pushed some changes. Can you integrate them?

File: stock_picking_invoice_link/views/stock_view.xml (hunk 10 above)
Comment (cubells) [reply stock_picking_invoice_link/views/stock_view.xml:33]: ok. You are right but this views is for stock_move model.

I see information is not shown so I have to correct this.

Comment (cubells) [reply stock_picking_invoice_link/views/stock_view.xml:33]: Information in these views are shown as you can see:

seleccio_331

seleccio_332

Do you think is necessary add a tree view for only show invoice line information?

Comment (chienandalu) [reply stock_picking_invoice_link/views/stock_view.xml:33]: Mmm, well, I'd expect at least a reference to wich invoice this line comes from and the ability to get to it.

**Comment by luismontalba** [comment]
Is this ready to review?

File: stock_picking_invoice_link/views/stock_view.xml (hunk 10 above)
Comment (pedrobaeza) [reply stock_picking_invoice_link/views/stock_view.xml:33]: @cubells here a tree view is needed for showing the minimal data:
- Invoice
- Product
- Quantity
- Price
- Taxes
- ...

File: stock_picking_invoice_link/models/stock_move.py (hunk 2 above)
Comment (pedrobaeza) [reply stock_picking_invoice_link/models/stock_move.py:14]: @cubells can you answer this question?

**Comment by pedrobaeza** [comment]
@luismontalba please re-review

**Review by luismontalba — APPROVED** [review]
Tested on runbot, working fine

**Review by pedrobaeza — APPROVED** [review]

**Pull request merged**